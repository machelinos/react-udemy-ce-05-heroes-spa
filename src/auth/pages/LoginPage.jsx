import { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context/AuthContext'

export const LoginPage = () => {
  const navigate = useNavigate()
  const { login } = useContext(AuthContext)

  const handleLogin = () => {
    login('Joy Marcelle')
    navigate('/', { replace: true })
  }
  return (
    <div className="container mt-4">
      <h1>Login Page</h1>
      <hr />

      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  )
}
