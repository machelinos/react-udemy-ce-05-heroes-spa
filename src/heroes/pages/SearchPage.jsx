import { useMemo } from 'react'
import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'
import { HeroCard } from '../components'
import { getHeroesByQuery } from '../helpers'
import { useForm } from '../../hooks/useForm'

export const SearchPage = () => {
  const navigate = useNavigate()

  const { search } = useLocation()
  const { q = '' } = queryString.parse(search)

  const { searchText, handleInputChange } = useForm({
    searchText: q,
  })

  const heroes = useMemo(() => getHeroesByQuery(q), [q])

  const handleSubmit = (event) => {
    event.preventDefault()
    navigate(`?q=${searchText}`)
  }

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <form onSubmit={handleSubmit} aria-label="form">
            <input
              className="form-control"
              type="text"
              placeholder="Search"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
            />

            <button type="submit" className="btn btn-outline-primary mt-2">
              Search
            </button>
          </form>
        </div>

        <div className="col-7">
          <h3>Results</h3>

          {heroes.length > 0 && (
            <div className="row row-cols-1">
              {heroes.map((hero) => (
                <HeroCard key={hero.id} {...hero} />
              ))}
            </div>
          )}

          {q.length <= 1 && (
            <div className="alert alert-primary">
              <p className="mb-0">Search a hero</p>
            </div>
          )}

          {q.length > 1 && heroes.length === 0 && (
            <div className="alert alert-danger">
              <p className="mb-0">
                No results for <strong>{q}</strong>
              </p>
            </div>
          )}
        </div>
      </div>
    </>
  )
}
