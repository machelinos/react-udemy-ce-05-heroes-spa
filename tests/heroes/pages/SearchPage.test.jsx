import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter, useNavigate } from 'react-router-dom'
import { SearchPage } from '../../../src/heroes/pages/SearchPage'

const mockedUseNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}))

describe('<SearchPage /> tests', () => {
  test('should load the component with the default values', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>,
    )

    expect(container).toMatchSnapshot()
  })

  test('should load the input with batman query', () => {
    const query = 'batman'
    render(
      <MemoryRouter initialEntries={[`?q=${query}`]}>
        <SearchPage />
      </MemoryRouter>,
    )
    const input = screen.getByRole('textbox')
    expect(input.value).toBe(query)
    expect(screen.getAllByText(/Batman/i)).toBeTruthy()
  })

  test('should show the error message if not heroes are found', () => {
    const query = 'asdaxa'
    render(
      <MemoryRouter initialEntries={[`?q=${query}`]}>
        <SearchPage />
      </MemoryRouter>,
    )
    expect(screen.getByText(/No results for/i)).toBeTruthy()
  })

  test('should navigate to the search query page', () => {
    const query = 'batman'
    render(
      <MemoryRouter initialEntries={[`?q=${query}`]}>
        <SearchPage />
      </MemoryRouter>,
    )
    const form = screen.getByLabelText('form')
    fireEvent.submit(form)
    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${query}`)
  })
})
