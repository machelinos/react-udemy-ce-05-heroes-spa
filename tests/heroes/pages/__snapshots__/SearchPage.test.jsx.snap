// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPage /> tests should load the component with the default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h3>
        Results
      </h3>
      <div
        class="alert alert-primary"
      >
        <p
          class="mb-0"
        >
          Search a hero
        </p>
      </div>
    </div>
  </div>
</div>
`;
